### PR TITLE
8334489: Add function os::used_memory

### PR DESCRIPTION
**Pull request by jdksjolen**
A function os::used_memory is needed for GC ergonomics.

A naïve implementation is:

```c++
julong os::used_memory() {
  return os::physical_memory() - os::available_memory();
}
```

This function does not work well in a containerized environment, as the amount of physical memory may change. To understand why, we must look under the hood.

```c++
julong os::used_memory() {
  return os::physical_memory() - os::available_memory();
}

// can be expanded into

julong os::used_memory() {
  // The os::physical_memory() call
  julong mem_physical1 = OSContainer::memory_limit_in_bytes();
  // The os::available_memory() call
  julong mem_used = OSContainer::memory_usage_in_bytes();
  julong mem_physical2 = OSContainer::memory_limit_in_bytes();
  
  // Uh-oh: mem_physical1 may differ from mem_physical2 at this point
  // That means that this number is wrong
  return mem_physical1 - (mem_physical2 - mem_used);
}
```

The fix is to expose OSContainer::memory_usage_in_bytes if it's available, as this call does not depend on OSContainer::memory_limit_in_bytes. The default implementation will use the naïve implementation. 

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334489](https://bugs.openjdk.org/browse/JDK-8334489): Add function os::used_memory (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [25beb82d](https://git.openjdk.org/jdk/pull/19772/files/25beb82d76bfb95f6f2beb67d9a143f490469bbd)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [c12539a4](https://git.openjdk.org/jdk/pull/19772/files/c12539a4d64064b0ea36cec7ac7032c71bc9ed36)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19772/head:pull/19772` \
`$ git checkout pull/19772`

Update a local copy of the PR: \
`$ git checkout pull/19772` \
`$ git pull https://git.openjdk.org/jdk.git pull/19772/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19772`

View PR using the GUI difftool: \
`$ git pr show -t 19772`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19772.diff">https://git.openjdk.org/jdk/pull/19772.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19772#issuecomment-2176450218)

### CODE REVIEW

**Review by fisk — APPROVED** [review]
Looks good to me! Thanks for fixing.

**Review by dholmes-ora — COMMENTED** [review]
I think in this case I'd prefer to have an ifdef LINUX in a single shared version in os.cpp rather than duplicating the function this way.

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -271,6 +271,16 @@ julong os::Linux::available_memory() {
   return avail_mem;
 }
 
+julong os::used_memory() {
+  if (OSContainer::is_containerized()) {
```
Comment (jerboaa) [inline src/hotspot/os/linux/os_linux.cpp:275]: I'm curious. What's the intended use case of this? https://bugs.openjdk.org/browse/JDK-8329758 perhaps? If this is getting used more broadly, https://bugs.openjdk.org/browse/JDK-8261242, would seem to get bumped in priority as that still affects physical Linux systems.

**Review by jerboaa — APPROVED** [review]
Seems OK to me from a container code related perspective. It would be good to know where this will be used though.

**Comment by tstuefe** [comment]
@jdksjolen You subtract available from physical. Why not free from physical? What is your definition of used? Used, as in "used by user processes", or used as in "used by anything on the machine, including kernel-side caches"?

I'd like to have good comments for the growing number of memory query APIs at some point. Maybe it would also be good to have a single compound API, platform-independent, akin to os::Linux::query_process_memory_info().

File: src/hotspot/os/linux/os_linux.cpp (hunk 1 above)
Comment (jdksjolen) [reply src/hotspot/os/linux/os_linux.cpp:275]: I think you hit the nail on the head regarding intended use case. It's probably important that that issue gets resolved, yes.

**Comment by jdksjolen** [comment]
@dholmes-ora 

> I think in this case I'd prefer to have an ifdef LINUX in a single shared version in os.cpp rather than duplicating the function this way.

That's nice, let's do that.

@tstuefe 

> You subtract available from physical. Why not free from physical?

`os::free_memory()` is just `available_memory()` on all of our platforms. I thought it was weird that we have two of them which are identical, so I just picked the one that is defined.

>What is your definition of used? Used, as in "used by user processes", or used as in "used by anything on the machine, including kernel-side caches"?

The latter, but machine can be replaced by "the container". To quote:

>[memory.current] Shows the total amount of memory currently being used by the cgroup and its descendants. It includes page cache, in-kernel data structures such as inodes, and network buffers.

That should be a viable explanation for the non-container case also.

**Comment by jerboaa** [comment]
> `os::free_memory()` is just `available_memory()` on all of our platforms. I thought it was weird that we have two of them which are identical, so I just picked the one that is defined.

That doesn't sound right. `available_memory()` is using `/proc/meminfo` on Linux (`MemAvailable:` key). While `free_memory()` uses `sysinfo`'s `freeram`. They are only identical for the container case where such a thing like `MemAvailable` is not available in the interface files.

TLDR; It's the same on all platforms except Linux.

File: src/hotspot/share/runtime/os.hpp [hunk 2]
```diff
@@ -336,6 +336,9 @@ class os: AllStatic {
   // than "free" memory (`MemFree` in `/proc/meminfo`) because Linux can free memory
   // aggressively (e.g. clear caches) so that it becomes available.
   static julong available_memory();
+#ifdef LINUX
+  static julong used_memory();
+#endif
```
Comment (jerboaa) [inline src/hotspot/share/runtime/os.hpp:341]: I don't understand. Now it's a linux only API? Wasn't the intention to have it available on all, but a shared impl for all except Linux, like in `os.cpp`?

Something like:
```
LINUX_ONLY(static julong pd_used_memory();)
...
julong os::used_memory() {
#ifdef LINUX
  return os::pd_used_memory();
#else
  return os::physical_memory() - os::available_memory();
#endif
}

**Comment by fisk** [comment]
> @jdksjolen You subtract available from physical. Why not free from physical? What is your definition of used? Used, as in "used by user processes", or used as in "used by anything on the machine, including kernel-side caches"?
> 
> 
> 
> I'd like to have good comments for the growing number of memory query APIs at some point. Maybe it would also be good to have a single compound API, platform-independent, akin to os::Linux::query_process_memory_info().

Just for some context... I'm working on automatic heap sizing for ZGC. Part of the heuristics want to reason about some sort of "memory pressure" on the machine, which is vaguely related to what portion of the machine or container memory is used vs not used. 

It struck me that with the existing APIs, I can't calculate what fraction of memory is neither used, free nor available. Because we only expose free and available which depend on the container size. And to calculate the fraction I would have to ask for the container size again. If the two reported sizes are inconsistent, you are bound to get nonsense values.

The used memory, i.e. memory that is not available, is the only number among these that stays invariant of the container size. And it's unfortunately the number we hide in the implementation of the others making it awkwardly not possible to get consistent fractions.

**Comment by jerboaa** [comment]
> Just for some context... I'm working on automatic heap sizing for ZGC. Part of the heuristics want to reason about some sort of "memory pressure" on the machine, which is vaguely related to what portion of the machine or container memory is used vs not used.

That work is not supposed to also adjust to changing memory limits **at JVM runtime** is it? I'm not aware that the JVM is yet ready to do that.

> Because we only expose free and available which depend on the container size. And to calculate the fraction I would have to ask for the container size again. If the two reported sizes are inconsistent, you are bound to get nonsense values.

Why would they be inconsistent? Change the limit at JVM runtime?

**Comment by fisk** [comment]
> > Just for some context... I'm working on automatic heap sizing for ZGC. Part of the heuristics want to reason about some sort of "memory pressure" on the machine, which is vaguely related to what portion of the machine or container memory is used vs not used.
> 
> 
> 
> That work is not supposed to also adjust to changing memory limits **at JVM runtime** is it? I'm not aware that the JVM is yet ready to do that.
> 
> 
> 
> > Because we only expose free and available which depend on the container size. And to calculate the fraction I would have to ask for the container size again. If the two reported sizes are inconsistent, you are bound to get nonsense values.
> 
> 
> 
> Why would they be inconsistent? Change the limit at JVM runtime?

Right. You can change the container size at JVM runtime, and I don't want to be caught off guard by that in my heuristics. There is an alpha kubernetes feature, InPlacePodVerticalScaling, for utilizing it already. So I much prefer doing the right thing and make this heuristic code robust in case someone starts using this.

File: src/hotspot/share/runtime/os.hpp (hunk 2 above)
Comment (fisk) [reply src/hotspot/share/runtime/os.hpp:341]: > I don't understand. Now it's a linux only API? Wasn't the intention to have it available on all, but a shared impl for all except Linux, like in `os.cpp`?
> 
> 
> 
> Something like:
> 
> ```
> 
> LINUX_ONLY(static julong pd_used_memory();)
> 
> ...
> 
> julong os::used_memory() {
> 
> #ifdef LINUX
> 
>   return os::pd_used_memory();
> 
> #else
> 
>   return os::physical_memory() - os::available_memory();
> 
> #endif
> 
> }

That was indeed the intention.

**Comment by tstuefe** [comment]
BTW does anyone know whats the deal with os::physical_memory vs os::Linux::physical_memory on Linux? The former is not cgroup-aware, the latter is. Why do we have two variants?

**Comment by jerboaa** [comment]
> BTW does anyone know whats the deal with os::physical_memory vs os::Linux::physical_memory on Linux? The former is not cgroup-aware, the latter is. Why do we have two variants?

For all intents and purposes `os::physical_memory()` should be used in VM code. `os::Linux::physical_memory()` is used in a) the container code to have an upper bound for memory limits (the v1 interface of cgroups doesn't have a `max` value like v2 has, for example). b) in some JFR events like `jdk.ContainerConfiguration` where host physical is explicitly queried in addition to the container limit. So we need both.

**Comment by jerboaa** [comment]
> > Why would they be inconsistent? Change the limit at JVM runtime?
> 
> Right. You can change the container size at JVM runtime, and I don't want to be caught off guard by that in my heuristics. There is an alpha kubernetes feature, InPlacePodVerticalScaling, for utilizing it already. So I much prefer doing the right thing and make this heuristic code robust in case someone starts using this.

OK. Making the JVM handle changing physical memory size at runtime will be a larger task, though.

File: src/hotspot/share/runtime/os.hpp (hunk 2 above)
Comment (jdksjolen) [reply src/hotspot/share/runtime/os.hpp:341]: Yeah that's way better, d'oh. I was too happy just getting to delete code.

**Review by tstuefe — APPROVED** [review]
+1

File: src/hotspot/share/runtime/os.cpp [hunk 3]
```diff
@@ -2064,6 +2064,19 @@ static void assert_nonempty_range(const char* addr, size_t bytes) {
...
 }
 
+julong os::used_memory() {
+#ifndef LINUX
+  if (OSContainer::is_containerized()) {
+    jlong mem_usage = OSContainer::memory_usage_in_bytes();
+    if (mem_usage > 0) {
+      return mem_usage;
+    }
+  }
```
Comment (jerboaa) [inline src/hotspot/share/runtime/os.cpp:2078]: I'm not sure we should introduce Linux specific things in the shared os code like that. It seems the model is to use `pd_<foo>` to delegate to platform specific code? @dholmes-ora thoughts?

**Review by jerboaa — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/os.cpp (hunk 3 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:2078]: I think that is okay here. Sometimes its the lesser of two evils.

I actually understood David's remark that way, too (https://github.com/openjdk/jdk/pull/19772#pullrequestreview-2127842338)

Comment (jerboaa) [reply src/hotspot/share/runtime/os.cpp:2078]: I understood the remark as "reduce the default implementation duplication". I.e. not have the same impl for all platforms. A pd_<foo> would just move the impl to `os/linux` rather than having it in `share/runtime`. But if that's fine, then I won't object, of course.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/os.cpp:2078]: I am okay with the proposed code. If it were larger then I'd want it hidden behind a pd function but this is okay.

**Review by dholmes-ora — APPROVED** [review]
Okay. Thanks

**Comment by jdksjolen** [comment]
Thank you!

/integrate
